### PR TITLE
feat(analysis): inter-procedural return-type propagation (Phase 8.2)

### PR DESCRIPTION
**Pull request by carlos-alm**
## Summary

- Adds `extractReturnTypeMapWalk` to the JS/TS extractor: scans every function/method definition and records its return type (confidence 1.0 from explicit TS annotations, 0.85 from `return new Constructor()`)
- Propagates return types intra-file: `const x = createUser()` → `x` gets type `User` in the typeMap; method chains like `getService().getRepo()` are resolved recursively up to 3 hops with confidence decaying 0.1 per hop (0.9 → 0.8 → 0.7)
- Cross-file propagation in `build-edges.ts`: unresolved call assignments are resolved against imported files' `returnTypeMap`s before both native and JS call-edge paths run
- Adds `analysis.typePropagationDepth: 3` to `DEFAULTS`/`CodegraphConfig` for future tunability

## Test plan

- [ ] `npm test` — all 2790 tests pass, 12 new tests covering: `returnTypeMap` extraction (annotations, inferred, method qualification, arrow fns), single-hop propagation, method-call propagation, chain resolution, confidence values, and edge cases (no-override, missing return info)
- [ ] `npx tsc --noEmit` — clean

Closes #[roadmap Phase 8.2]

### CODE REVIEW

File: src/infrastructure/config.ts [hunk 1]
```diff
@@ -79,6 +79,7 @@ export const DEFAULTS = {
     briefImporterDepth: 5,
     briefHighRiskCallers: 10,
     briefMediumRiskCallers: 3,
+    typePropagationDepth: 3,
```
[the remark this answers is not in the document]
Comment (carlos-alm) [reply src/infrastructure/config.ts:85]: Fixed — the config field now carries a TODO comment pointing to MAX_PROPAGATION_DEPTH and explaining this will be wired through when config is threaded into extractSymbols (Phase 8.3). The comment explicitly notes the current code uses the hardcoded constant, so it won't be mistaken for a live control knob.

File: src/extractors/javascript.ts [hunk 2]
```diff
@@ -1114,22 +1138,221 @@ function extractNewExprTypeName(newExprNode: TreeSitterNode): string | null {
...
+          const fnName = currentClass ? `${currentClass}.${nameN.text}` : nameN.text;
+          storeReturnType(valueN, fnName, returnTypeMap);
+        }
+      }
+    }
+
+    for (let i = 0; i < node.childCount; i++) {
+      walk(node.child(i)!, depth + 1, currentClass);
+    }
+  }
```
[the remark this answers is not in the document]
Comment (carlos-alm) [reply src/extractors/javascript.ts:1207]: Fixed — when recursing into the children of function_declaration and method_definition nodes, the walk now passes null as currentClass. This prevents any function_declaration or arrow-function variable_declarator nested inside a method body from inheriting the enclosing class name. Added early return statements after the null-class recursive calls to avoid double-visiting.

File: src/extractors/javascript.ts [hunk 3]
```diff
@@ -1114,22 +1138,221 @@ function extractNewExprTypeName(newExprNode: TreeSitterNode): string | null {
...
+  returnTypeMap: Map<string, TypeMapEntry>,
+): void {
+  const returnTypeNode = fnNode.childForFieldName('return_type');
+  if (returnTypeNode) {
+    const typeName = extractSimpleTypeName(returnTypeNode);
+    if (typeName) {
+      const existing = returnTypeMap.get(fnName);
+      if (!existing || 1.0 > existing.confidence)
+        returnTypeMap.set(fnName, { type: typeName, confidence: 1.0 });
+      return;
```
[the remark this answers is not in the document]
Comment (carlos-alm) [reply src/extractors/javascript.ts:1224]: Fixed — changed `1.0 > existing.confidence` to `existing.confidence < 1.0`. This is semantically identical in effect but clearly expresses the intent: "overwrite unless already at maximum confidence", making it obvious the first annotation at 1.0 wins for duplicate function names.

File: src/domain/graph/builder/stages/build-edges.ts [hunk 4]
```diff
@@ -846,6 +905,10 @@ export async function buildEdges(ctx: PipelineContext): Promise<void> {
       buildImportEdges(ctx, getNodeIdStmt, allEdgeRows);
     }
 
+    // Phase 8.2: Augment typeMaps with cross-file return-type propagation before
+    // call-edge building so both native and JS paths see the enriched typeMaps.
+    propagateReturnTypesAcrossFiles(ctx.fileSymbols, ctx, ctx.rootDir);
```
[the remark this answers is not in the document]
Comment (carlos-alm) [reply src/domain/graph/builder/stages/build-edges.ts]: Fixed — moved `propagateReturnTypesAcrossFiles` to just before `computeEdgesTx()` is called (outside the transaction). Since it performs no DB I/O, it needs no transactional protection; running it inside was a risk for leaving ctx.fileSymbols in a partial mutated state if the transaction rolled back.

**Comment by carlos-alm** [comment]
Addressed the review feedback:

**Fixed:**
1. **currentClass leak (P1 bug)** — `function_declaration` and `method_definition` handlers now recurse with `null` as `currentClass`, preventing nested function declarations inside method bodies from being keyed under the enclosing class name. Added early `return` to avoid double-visiting. Commit: fc9eb64.

2. **Annotation guard condition** — Changed `1.0 > existing.confidence` to `existing.confidence < 1.0`. Semantically equivalent but clearly expresses the intent. Commit: fc9eb64.

3. **Transaction placement** — Moved `propagateReturnTypesAcrossFiles` to immediately before `computeEdgesTx()` (outside the transaction), as it performs no DB I/O and should not risk leaving in-memory typeMap in a partial state on rollback. Commit: f9b306a.

4. **Hardcoded `- 0.1`** — Exported `PROPAGATION_HOP_PENALTY` from `javascript.ts` and imported it in `build-edges.ts` to replace the literal. The cross-file and intra-file paths now share the same constant. Commit: f9b306a.

5. **`typePropagationDepth` config field** — Added a TODO comment on the config entry and on `CodegraphConfig.analysis` type pointing to the hardcoded `MAX_PROPAGATION_DEPTH` constant and noting Phase 8.3 is where the wiring will happen. Commit: f9b306a.

**Tracked as issue (out of scope for this PR):**
6. **Native engine parity gap** — `extractReturnTypeMapWalk` + `recordCallAssignment` are JS/TS WASM-only. Implementing equivalent logic in the Rust extractor is a separate feature. Filed as #1280 to ensure it is tracked per CLAUDE.md requirement that both engines produce identical results.

**Comment by carlos-alm** [comment]
@greptileai

**Comment by carlos-alm** [comment]
@claude
